src/chaboche.jl

```diff
@@ -99,49 +167,89 @@ function integrate_material!(material::Chaboche)
                                           cumeq = cumeq,
                                           jacobian = jacobian)
     material.variables_new = variables_new
+    return nothing
 end
 
-function create_nonlinear_system_of_equations(material::Chaboche)
+"""
+    create_nonlinear_system_of_equations(material::Chaboche)
+
+Create and return an instance of the equation system for the incremental form of
+the evolution equations of the Chaboche material.
+
+Used internally for computing the plastic contribution in `integrate_material!`.
+
+The input `material` represents the problem state at the end of the previous
+timestep. The created equation system will hold its own copy of that state.
+
+The equation system is represented as a mutating function `g!` that computes the
+residual:
+
+```julia
+    g!(F::V, x::V) where V <: AbstractVector{<:Real}
+```
+
+Both `F` (output) and `x` (input) are length-19 vectors containing
+[sigma, R, X1, X2], in that order. The tensor quantities sigma, X1,
+X2 are encoded in Voigt format.
+
+The function `g!` is intended to be handed over to `nlsolve`.
+"""
+function create_nonlinear_system_of_equations(material::Chaboche{T}) where T <: Real
```

Comment:
I started thinking that maybe this unpacking, that is done twice now, is in vain. Would it make sense to move the `g!` function definition directly to the `integrate_material!`? Only the accepted `stress` is corrupted in the `integrate_material!` due to the elastic trial, but that can be worked around easily. What do you think?

Reply:
Code is read much more often than it is written, and [reading code is harder than writing it](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/), so almost always, clarity trumps speed.

However, numerics has the peculiar characteristic that, unlike in general software engineering, in numerics it's often possible to see *a priori* where performance bottlenecks will arise. From the viewpoint of the FEM solution process, this code obviously runs in a tight loop. Speed may be a concern here.

One thing that comes to mind though, is that Julia's compiler is fairly advanced - are we sure it won't optimize away the extra unpacking automatically? The documentation explicitly states that the compiler may inline functions wherever it sees fit, and in our case, `integrate_material!` is the only call site of the function `create_nonlinear_system_of_equations`, which in turn is private to `ChabocheModule`. It's an obvious candidate for inlining. Of course, the compiler must still respect the rules of lexical scope, so any (mutable) variables created in the callee must be kept distinct from those of the caller.

What I'm getting at is that if other properties of the material are only read (not written to), only one copy is really needed for them, and only `stress` needs to exist in duplicate. So if the function is inlined, it is in theory possible to refactor the result as unpack once, then copy the (already unpacked) stress variable, and for anything that is only read, just use a single copy. (Or even, if some property is only ever read once, just use a pointer to its memory address in the original data structure, and read it via that, no copy needed. But this is getting pretty close to the metal already.)

Still, as D. Knuth famously stated, [*Premature optimization is the root of all evil*](http://wiki.c2.com/?PrematureOptimization). So to be sure, I think we should profile this both ways to see whether we get a speed increase that is significant enough to pay for the decrease in clarity.

To be done later?